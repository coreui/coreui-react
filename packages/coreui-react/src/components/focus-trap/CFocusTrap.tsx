import React, { FC, ReactElement, cloneElement, useEffect, useRef } from 'react'
import { mergeRefs, focusableChildren } from './utils'

export interface CFocusTrapProps {
  /**
   * Controls whether the focus trap is active or inactive.
   * When `true`, focus will be trapped within the child element.
   * When `false`, normal focus behavior is restored.
   *
   * @default true
   */
  active?: boolean

  /**
   * Additional container elements to include in the focus trap.
   * Useful for floating elements like tooltips or popovers that are
   * rendered outside the main container but should be part of the trap.
   */
  additionalContainer?: React.RefObject<HTMLElement | null>

  /**
   * Single React element that renders a DOM node and forwards refs properly.
   * The focus trap will be applied to this element and all its focusable descendants.
   *
   * Requirements:
   * - Must be a single ReactElement (not an array or fragment)
   * - Must forward the ref to a DOM element
   * - Should contain focusable elements for proper trap behavior
   */
  children: ReactElement

  /**
   * Controls whether to focus the first selectable element or the container itself.
   * When `true`, focuses the first tabbable element within the container.
   * When `false`, focuses the container element directly.
   *
   * This is useful for containers that should receive focus themselves,
   * such as scrollable regions or custom interactive components.
   *
   * @default false
   */
  focusFirstElement?: boolean

  /**
   * Callback function invoked when the focus trap becomes active.
   * Useful for triggering additional accessibility announcements or analytics.
   */
  onActivate?: () => void

  /**
   * Callback function invoked when the focus trap is deactivated.
   * Can be used for cleanup, analytics, or triggering state changes.
   */
  onDeactivate?: () => void

  /**
   * Automatically restores focus to the previously focused element when the trap is deactivated.
   * This is crucial for accessibility as it maintains the user's place in the document
   * when returning from modal dialogs or overlay components.
   *
   * Recommended to be `true` for modal dialogs and popover components.
   *
   * @default true
   */
  restoreFocus?: boolean
}

export const CFocusTrap: FC<CFocusTrapProps> = ({
  active = true,
  additionalContainer,
  children,
  focusFirstElement = false,
  onActivate,
  onDeactivate,
  restoreFocus = true,
}) => {
  const containerRef = useRef<HTMLElement | null>(null)
  const prevFocusedRef = useRef<HTMLElement | null>(null)
  const isActiveRef = useRef<boolean>(false)
  const lastTabNavDirectionRef = useRef<'forward' | 'backward'>('forward')
  const tabEventSourceRef = useRef<HTMLElement | null>(null)

  useEffect(() => {
    const container = containerRef.current
    const _additionalContainer = additionalContainer?.current || null

    if (!active || !container) {
      if (isActiveRef.current) {
        // Deactivate cleanup
        if (restoreFocus && prevFocusedRef.current?.isConnected) {
          prevFocusedRef.current.focus({ preventScroll: true })
        }

        onDeactivate?.()
        isActiveRef.current = false
        prevFocusedRef.current = null
      }

      return
    }

    // Remember focused element BEFORE we move focus into the trap
    prevFocusedRef.current = document.activeElement as HTMLElement | null

    // Activating…
    isActiveRef.current = true

    // Set initial focus
    if (focusFirstElement) {
      const elements = focusableChildren(container)
      if (elements.length > 0) {
        elements[0].focus({ preventScroll: true })
      } else {
        // Fallback to container if no focusable elements
        container.focus({ preventScroll: true })
      }
    } else {
      container.focus({ preventScroll: true })
    }

    onActivate?.()

    const handleFocusIn = (event: FocusEvent) => {
      // Only handle focus events from tab navigation
      if (containerRef.current !== tabEventSourceRef.current) {
        return
      }

      const target = event.target as Node

      // Allow focus within container
      if (target === document || target === container || container.contains(target)) {
        return
      }

      // Allow focus within additional elements
      if (
        _additionalContainer &&
        (target === _additionalContainer || _additionalContainer.contains(target))
      ) {
        return
      }

      // Focus escaped, bring it back
      const elements = focusableChildren(container)

      if (elements.length === 0) {
        container.focus({ preventScroll: true })
      } else if (lastTabNavDirectionRef.current === 'backward') {
        elements.at(-1)?.focus({ preventScroll: true })
      } else {
        elements[0].focus({ preventScroll: true })
      }
    }

    const handleKeyDown = (event: KeyboardEvent) => {
      if (event.key !== 'Tab') {
        return
      }

      tabEventSourceRef.current = container
      lastTabNavDirectionRef.current = event.shiftKey ? 'backward' : 'forward'

      if (!_additionalContainer) {
        return
      }

      const containerElements = focusableChildren(container)
      const additionalElements = focusableChildren(_additionalContainer)

      if (containerElements.length === 0 && additionalElements.length === 0) {
        // No focusable elements, prevent tab
        event.preventDefault()
        return
      }

      const activeElement = document.activeElement as HTMLElement
      const isInContainer = containerElements.includes(activeElement)
      const isInAdditional = additionalElements.includes(activeElement)

      // Handle tab navigation between container and additional elements
      if (isInContainer) {
        const index = containerElements.indexOf(activeElement)

        if (
          !event.shiftKey &&
          index === containerElements.length - 1 &&
          additionalElements.length > 0
        ) {
          // Tab forward from last container element to first additional element
          event.preventDefault()
          additionalElements[0].focus({ preventScroll: true })
        } else if (event.shiftKey && index === 0 && additionalElements.length > 0) {
          // Tab backward from first container element to last additional element
          event.preventDefault()
          additionalElements.at(-1)?.focus({ preventScroll: true })
        }
      } else if (isInAdditional) {
        const index = additionalElements.indexOf(activeElement)

        if (
          !event.shiftKey &&
          index === additionalElements.length - 1 &&
          containerElements.length > 0
        ) {
          // Tab forward from last additional element to first container element
          event.preventDefault()
          containerElements[0].focus({ preventScroll: true })
        } else if (event.shiftKey && index === 0 && containerElements.length > 0) {
          // Tab backward from first additional element to last container element
          event.preventDefault()
          containerElements.at(-1)?.focus({ preventScroll: true })
        }
      }
    }

    // Add event listeners
    container.addEventListener('keydown', handleKeyDown, true)
    if (_additionalContainer) {
      _additionalContainer.addEventListener('keydown', handleKeyDown, true)
    }
    document.addEventListener('focusin', handleFocusIn, true)

    // Cleanup function
    return () => {
      container.removeEventListener('keydown', handleKeyDown, true)
      if (_additionalContainer) {
        _additionalContainer.removeEventListener('keydown', handleKeyDown, true)
      }
      document.removeEventListener('focusin', handleFocusIn, true)

      // On unmount (also considered deactivation)
      if (restoreFocus && prevFocusedRef.current?.isConnected) {
        prevFocusedRef.current.focus({ preventScroll: true })
      }

      if (isActiveRef.current) {
        onDeactivate?.()
        isActiveRef.current = false
      }

      prevFocusedRef.current = null
    }
  }, [active, additionalContainer, focusFirstElement, onActivate, onDeactivate, restoreFocus])

  // Attach our ref to the ONLY child — no extra wrappers
  const onlyChild = React.Children.only(children)
  const childRef = (onlyChild as React.ReactElement & { ref?: React.Ref<HTMLElement> }).ref
  const mergedRef = mergeRefs(childRef, (node: HTMLElement | null) => {
    containerRef.current = node
  })

  return cloneElement(onlyChild, { ref: mergedRef } as { ref: React.Ref<HTMLElement> })
}
