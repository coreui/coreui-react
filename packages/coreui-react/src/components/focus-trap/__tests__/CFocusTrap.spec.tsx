import * as React from 'react'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import { CFocusTrap } from '../CFocusTrap'

// Helper function to create a test component with focusable elements
interface TestComponentProps {
  children?: React.ReactNode
  [key: string]: any
}

const TestComponent = ({ children, ...props }: TestComponentProps) => (
  <CFocusTrap {...props}>
    <div data-testid="container">
      <button data-testid="first-button">First</button>
      <input data-testid="input" type="text" placeholder="Input field" />
      <a href="#" data-testid="link">
        Link
      </a>
      <button data-testid="last-button">Last</button>
      {children}
    </div>
  </CFocusTrap>
)

describe('CFocusTrap', () => {
  beforeEach(() => {
    // Reset document focus before each test
    document.body.focus()
  })

  test('loads and displays CFocusTrap component', () => {
    const { container } = render(
      <CFocusTrap>
        <div data-testid="test-content">Test Content</div>
      </CFocusTrap>
    )
    expect(container).toMatchSnapshot()
    expect(screen.getByTestId('test-content')).toBeInTheDocument()
  })

  test('CFocusTrap with custom props', () => {
    const onActivate = jest.fn()
    const onDeactivate = jest.fn()

    const { container } = render(
      <CFocusTrap
        active={true}
        restoreFocus={false}
        focusFirstElement={false}
        onActivate={onActivate}
        onDeactivate={onDeactivate}
      >
        <div data-testid="custom-container">Custom Content</div>
      </CFocusTrap>
    )

    expect(container).toMatchSnapshot()
    expect(onActivate).toHaveBeenCalledTimes(1)
  })

  test('focuses container when focusFirstElement is false (default)', () => {
    const mockFocus = jest.fn()
    const originalFocus = HTMLElement.prototype.focus
    HTMLElement.prototype.focus = mockFocus

    render(<TestComponent active={true} />)

    const container = screen.getByTestId('container')
    expect(container).toBeInTheDocument()
    expect(mockFocus).toHaveBeenCalledWith({ preventScroll: true })

    HTMLElement.prototype.focus = originalFocus
  })

  test('does not trap focus when active is false', () => {
    render(<TestComponent active={false} />)

    // Focus should not be moved to any element
    expect(screen.getByTestId('container')).not.toHaveFocus()
    expect(screen.getByTestId('first-button')).not.toHaveFocus()
  })

  test('handles container with no tabbable elements', () => {
    const mockFocus = jest.fn()
    const originalFocus = HTMLElement.prototype.focus
    HTMLElement.prototype.focus = mockFocus

    render(
      <CFocusTrap active={true}>
        <div data-testid="empty-container">No focusable elements</div>
      </CFocusTrap>
    )

    const container = screen.getByTestId('empty-container')
    expect(container).toBeInTheDocument()
    expect(mockFocus).toHaveBeenCalledWith({ preventScroll: true })

    HTMLElement.prototype.focus = originalFocus
  })

  test('calls onActivate callback when trap becomes active', () => {
    const onActivate = jest.fn()

    const { rerender } = render(<TestComponent active={false} onActivate={onActivate} />)
    expect(onActivate).not.toHaveBeenCalled()

    // Re-render with active=true
    rerender(<TestComponent active={true} onActivate={onActivate} />)
    expect(onActivate).toHaveBeenCalledTimes(1)
  })

  test('calls onDeactivate callback when trap becomes inactive', () => {
    const onDeactivate = jest.fn()

    const { rerender } = render(<TestComponent active={true} onDeactivate={onDeactivate} />)
    expect(onDeactivate).not.toHaveBeenCalled()

    // Deactivate the trap
    rerender(<TestComponent active={false} onDeactivate={onDeactivate} />)
    expect(onDeactivate).toHaveBeenCalledTimes(1)
  })

  test('cleans up event listeners on unmount', () => {
    const removeEventListenerSpy = jest.spyOn(document, 'removeEventListener')

    const { unmount } = render(<TestComponent active={true} />)

    unmount()

    expect(removeEventListenerSpy).toHaveBeenCalledWith('focusin', expect.any(Function), true)

    removeEventListenerSpy.mockRestore()
  })

  test('focuses first element when focusFirstElement is true', () => {
    const mockFocus = jest.fn()
    const originalFocus = HTMLElement.prototype.focus
    HTMLElement.prototype.focus = mockFocus

    render(<TestComponent active={true} focusFirstElement={true} />)

    const firstButton = screen.getByTestId('first-button')
    expect(firstButton).toBeInTheDocument()
    expect(mockFocus).toHaveBeenCalledWith({ preventScroll: true })

    HTMLElement.prototype.focus = originalFocus
  })

  test('works with additionalContainer prop', () => {
    const additionalRef = React.createRef<HTMLDivElement>()

    render(
      <div>
        <CFocusTrap active={true} additionalContainer={additionalRef}>
          <div data-testid="main-container">
            <button data-testid="main-button">Main Button</button>
          </div>
        </CFocusTrap>
        <div ref={additionalRef} data-testid="additional-container">
          <button data-testid="additional-button">Additional Button</button>
        </div>
      </div>
    )

    const mainContainer = screen.getByTestId('main-container')
    const additionalContainer = screen.getByTestId('additional-container')

    expect(mainContainer).toBeInTheDocument()
    expect(additionalContainer).toBeInTheDocument()
  })

  test('restores focus when restoreFocus is true', () => {
    // Mock document.activeElement to simulate a focused element
    const focusButton = document.createElement('button')
    focusButton.dataset.testid = 'focus-button'
    document.body.appendChild(focusButton)

    const mockFocus = jest.fn()
    focusButton.focus = mockFocus

    // Mock document.activeElement
    Object.defineProperty(document, 'activeElement', {
      value: focusButton,
      writable: true,
      configurable: true,
    })

    const { rerender } = render(<TestComponent active={false} restoreFocus={true} />)

    // Activate the trap (should store the previous focused element)
    rerender(<TestComponent active={true} restoreFocus={true} />)

    // Deactivate the trap (should restore focus)
    rerender(<TestComponent active={false} restoreFocus={true} />)

    // Verify focus restoration was attempted
    expect(mockFocus).toHaveBeenCalledWith({ preventScroll: true })

    focusButton.remove()
  })

  test('does not restore focus when restoreFocus is false', () => {
    // Mock document.activeElement to simulate a focused element
    const focusButton = document.createElement('button')
    focusButton.dataset.testid = 'focus-button'
    document.body.appendChild(focusButton)

    const mockFocus = jest.fn()
    focusButton.focus = mockFocus

    // Mock document.activeElement
    Object.defineProperty(document, 'activeElement', {
      value: focusButton,
      writable: true,
      configurable: true,
    })

    const { rerender } = render(<TestComponent active={false} restoreFocus={false} />)

    // Activate the trap (should store the previous focused element)
    rerender(<TestComponent active={true} restoreFocus={false} />)

    // Deactivate the trap (should NOT restore focus)
    rerender(<TestComponent active={false} restoreFocus={false} />)

    // Verify focus restoration was not attempted
    expect(mockFocus).not.toHaveBeenCalledWith({ preventScroll: true })

    focusButton.remove()
  })
})
