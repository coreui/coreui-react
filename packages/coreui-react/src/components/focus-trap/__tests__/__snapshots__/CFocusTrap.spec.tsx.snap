// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CFocusTrap CFocusTrap with custom props 1`] = `
<div>
  <div
    data-testid="custom-container"
  >
    Custom Content
  </div>
</div>
`;

exports[`CFocusTrap loads and displays CFocusTrap component 1`] = `
<div>
  <div
    data-testid="test-content"
  >
    Test Content
  </div>
</div>
`;
